src/core/server/saved_objects/service/lib/create_repository.ts

```diff
@@ -0,0 +1,65 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { CallAPIOptions } from 'src/core/server/elasticsearch';
+import { Config } from 'src/core/server/config';
+import { retryCallCluster } from '../../../elasticsearch/retry_call_cluster';
+import { KibanaMigrator } from '../../migrations';
+import { SavedObjectsSchema } from '../../schema';
+import { getRootPropertiesObjects } from '../../mappings';
+import { SavedObjectsSerializer } from '../../serialization';
+import { SavedObjectsRepository } from '.';
+
+export const createRepository = (
```

Comment:
The repository factory has some important logic that's hard to test if it's embedded in the repository constructor. I created a separate file for this factory to make it easier to test as it allows me to mock the whole repository module and make assertions against the mocked constructor.

Reply:
Maybe this would be more clear as a `public static` function on the SavedObjectRepository class with a `private` constructor? I know that would make the mocking a bit more complicated though.

Reply:
Yeah that's what I had at first but it seems like it's impossible to mock or spy on constructor functions unless you use an import mock. I couldn't find any documentation on jest, but this thread from sinon explains the problem: https://github.com/sinonjs/sinon/issues/831#issuecomment-198081263 So I like the code better, but ultimately choose better testability.

Reply:
I tried the "minimal dependency injection" approach. The testability "leaks" out of the abstraction in that you have to pass in a constructor for your factory function, but it otherwise works. I'm not sure it's clearer but I can implement it?

```diff
diff --git a/src/core/server/saved_objects/service/lib/create_repository.test.ts b/src/core/server/saved_objects/service/lib/create_repository.test.ts
index d40a5d04dc..10825354a5 100644
--- a/src/core/server/saved_objects/service/lib/create_repository.test.ts
+++ b/src/core/server/saved_objects/service/lib/create_repository.test.ts
@@ -24,6 +24,8 @@ import { Config } from 'src/core/server/config';
 import { createRepository } from './create_repository';
 jest.mock('./repository');
 
+const { SavedObjectsRepository: originalRepository } = jest.requireActual('./repository');
+
 describe('#createRepository', () => {
   const callAdminCluster = jest.fn();
   const schema = new SavedObjectsSchema({
@@ -64,7 +66,7 @@ describe('#createRepository', () => {
 
   it('should not allow a repository with an undefined type', () => {
     try {
-      createRepository(
+      originalRepository.createFactory(
         (migrator as unknown) as KibanaMigrator,
         schema,
         {} as Config,
@@ -79,13 +81,15 @@ describe('#createRepository', () => {
     }
   });
 
-  it('should create a repository without hidden types', () => {
-    const repository = createRepository(
+  it.only('should create a repository without hidden types', () => {
+    const repository = originalRepository.createFactory(
       (migrator as unknown) as KibanaMigrator,
       schema,
       {} as Config,
       '.kibana-test',
-      callAdminCluster
+      callAdminCluster,
+      [],
+      SavedObjectsRepository
     );
     expect(repository).toBeDefined();
     expect(RepositoryConstructor.mock.calls[0][0].allowedTypes).toMatchInlineSnapshot(`
diff --git a/src/core/server/saved_objects/service/lib/repository.ts b/src/core/server/saved_objects/service/lib/repository.ts
index d04cc7e938..612fcb0882 100644
--- a/src/core/server/saved_objects/service/lib/repository.ts
+++ b/src/core/server/saved_objects/service/lib/repository.ts
@@ -19,6 +19,8 @@
 
 import { omit } from 'lodash';
 import { CallCluster } from 'src/legacy/core_plugins/elasticsearch';
+import { APICaller } from 'src/core/server/elasticsearch';
+import { retryCallCluster } from '../../../elasticsearch/retry_call_cluster';
 import { getRootPropertiesObjects, IndexMapping } from '../../mappings';
 import { getSearchDsl } from './search_dsl';
 import { includedFields } from './included_fields';
@@ -123,8 +125,43 @@ export class SavedObjectsRepository {
   private _unwrappedCallCluster: CallCluster;
   private _serializer: SavedObjectsSerializer;
 
+  public static createFactory(
+    migrator: KibanaMigrator,
+    schema: SavedObjectsSchema,
+    config: Config,
+    indexName: string,
+    callCluster: APICaller,
+    extraTypes: string[] = [],
+    injectedConstructor: any
+  ) {
+    const mappings = migrator.getActiveMappings();
+    const allTypes = Object.keys(getRootPropertiesObjects(mappings));
+    const serializer = new SavedObjectsSerializer(schema);
+    const visibleTypes = allTypes.filter(type => !schema.isHiddenType(type));
+
+    const missingTypeMappings = extraTypes.filter(type => !allTypes.includes(type));
+    if (missingTypeMappings.length > 0) {
+      throw new Error(
+        `Missing mappings for saved objects types: '${missingTypeMappings.join(', ')}'`
+      );
+    }
+
+    const allowedTypes = [...new Set(visibleTypes.concat(extraTypes))];
+
+    return new injectedConstructor({
+      index: indexName,
+      config,
+      migrator,
+      mappings,
+      schema,
+      serializer,
+      allowedTypes,
+      callCluster: retryCallCluster(callCluster),
+    });
+  }
+
   /** @internal */
-  constructor(options: SavedObjectsRepositoryOptions) {
+  private constructor(options: SavedObjectsRepositoryOptions) {
     const {
       index,
       config,
```

Reply:
I do like the private constructor though.

Reply:
Nice! With a default argument for `injectedConstructor`, I think the API would be transparent to consumers, no?

The default could be: `(options: SavedObjectsRepositoryOptions) => new SavedObjectsRepository(options)`

Reply:
yes, consumers won't notice. I'll do the factory / constructor refactoring in a separate PR